[Java] Implement concept exercise conditionals-ternary (#2265)
exercism/v3: languages/java/exercises/concept/conditionals-ternary/.docs/after.md

> _ternary operators_ can be thought of as being _if-else_ inline. It is a good example that most of the times you have different ways of achieving the same result.

Reviewer: The `after.md` document should contain all information that is introduced in the .docs/introduction.md document. See [the docs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md#docsaftermd). It doesn't necessarily have to use the exact same wording, but it feels like this document is missing the introduction bit.

Author: Oh, i didn't remember that, i will modify it

Reviewer: Great!

So, how to decide between _if-else_ and _ternary_ ? Well, _Ternary_ is used in simple cenarios, where you just need to return a value based on a condition and no extra computation is needed, use _if-else_ for everthing else, like nested conditions, big expressions and when more than one line is needed to decide the return value.

And try to never nest _ternary operators_ as the expression becomes illegible, prefer nested if's to handle those cases

```java

// hard to read
int value = expr1 ? expr2 ? val1 : expr3 ? val2 : val3 : val4

> // easier to read
> if (expr1){
>     if (expr2){
>         return val1;
>     } else{
>         if (expr3) {
>             return val2;
>         } else {
>             return val3;
>         }
>     }
> } else {
>     return val4;
> }

Reviewer: I'm no Java expert, but I find this not easier to read :)

```suggestion
// easier to read
if (expr1){
    if (expr2){
        return val1;
    } else{
        if (expr3) {
            return val2;
        } else {
            return val3;
        }
    }
} else {
    return val4;
}
```

Author: Unfortunately i understand you, i think most of this is due to the brackets.
I think a shallow nested if will do the trick

Reviewer: 👍

Author: ```java
// hard to read
int value = expr1 ? expr2 && expr3 ? val1 : (val2 + val3) : val4;

// easier to read
if (expr1){

    if (expr2 && expr3){

        return val1;
    } else {

        return val2 + val3;
    }

} else {
    
    return val4;
}

 ```

## Ternary Operator

> The _ternary operators_ can be thought of as being a compact version of _if-else_. It's usually used in (but not restricted to) return statements, it needs just one single to make the decision, returning the left value if the expression is `true` and the right value if `false`.

Reviewer: I think a word might be missing here.

```suggestion
The _ternary operators_ can be thought of as being a compact version of _if-else_. It's usually used in (but not restricted to) return statements, it needs just one single line to make the decision, returning the left value if the expression is `true` and the right value if `false`.
```

> A lot of simple _if/else_ expressions can be simplified using _ternary operators_

Reviewer: ```suggestion
A lot of simple _if/else_ expressions can be simplified using _ternary operators_.
```

```java

> 

Reviewer: ```suggestion
```

// this ternary statement
return 5 > 4 ? true : false;

// is equivalent to this if/else
if ( 5 > 4 ) {
    return true;
} else {
    return false;
}
```

> So, how to decide between _if-else_ and _ternary_ ? Well, _Ternary_ is used in simple cenarios, where you just need to return a value based on a condition and no extra computation is needed, use _if-else_ for everthing else, like nested conditions, big expressions and when more than one line is needed to decide the return value.

Reviewer: ```suggestion
So, how to decide between _if-else_ and _ternary_ ? Well, _ternary operators_ are used in simple scenarios, where you just need to return a value based on a condition and no extra computation is needed. Use an _if-else_ for everthing else, like nested conditions, big expressions and when more than one line is needed to decide the return value.
```

Reviewer: This section is great! It is sections like this that are perfect for an after.md document.

> And don't abuse of nested _ternary operators_ as the expression becomes confusing, prefer nested if's to handle those cases.

Reviewer: Maybe rephrase this slightly? Abuse sounds a bit strong IMHO.

```suggestion
While you can nest _ternary operators_, the code often becomes hard to read. In these cases, nested if's are preferred.
```

Author: of course, hahha

```java

> 

Reviewer: ```suggestion
```

// hard to read
int value = expr1 ? expr2 && expr3 ? val1 : (val2 + val3) : expr4;

// easier to read

> if (expr1){
> 
>     if (expr2 && expr3){
> 
>         return val1;
>     } else {
> 
>         return val2 + val3;
>     }
> 
> } else {
> 
>     return val4;
> }
> 

Reviewer: ```suggestion
if (expr1){
    if (expr2 && expr3){
        return val1;
    } else {
        return val2 + val3;
    }
} else {
    return val4;
}
```

Reviewer: I don't know if this is idiomatic in Java, but you could consider promoting an early return variant:

```java
if (!expr1){
    return val4;
}

if (expr2 && expr3){
    return val1;
}

return val2 + val3;
```

Author: I don't usually see this, but i think the first expression will already be enough
Do you think it's clear to the student why we can remove the _else's_?

Reviewer: Okay, let's go with just the first version.

```

> _Ternary_ and _if/else_ statements are a good example that you have different ways of achieving the same result when programming.

Reviewer: ```suggestion
_Ternary operators_ and _if/else_ statements are a good example that you have different ways of achieving the same result when programming.
```
